src/Common/HashTable/TimeBucketHashTable.h

```diff
@@ -438,4 +454,45 @@ class TimeBucketHashTable : private boost::noncopyable, protected Hash /// empty
 
         return buckets;
     }
+
+    bool isUpdatedBucket(Int64 bucket_) const
+    {
+        auto it = bucket_updated_flags.find(bucket_);
+        if (it != bucket_updated_flags.end())
+            return it->second;
+
+        return false;
+    }
+
+    void resetUpdated(Int64 bucket_)
+    {
+        auto it = bucket_updated_flags.find(bucket_);
+        if (it != bucket_updated_flags.end())
+            it->second = false;
+    }
+
+    void writeBucketUpdatedFlags(DB::WriteBuffer & wb) const
+    {
+        DB::writeVarUInt(bucket_updated_flags.size(), wb);
+        for (const auto & [bucket, updated] : bucket_updated_flags)
+        {
+            DB::writeIntBinary(bucket, wb);
+            DB::writeBoolText(updated, wb);
+        }
+    }
+
+    void readBucketUpdatedFlags(DB::ReadBuffer & rb)
+    {
+        size_t size = 0;
+        DB::readVarUInt(size, rb);
+        bucket_updated_flags.clear();
```

Comment:
we shall assert bucket_updated_flags is empty

Reply:
This method is called after deserialization of TwoLevelHashMap. so it always not empty.

Reply:
We are doing the deserialization here, if before the deserialization, it already has data, there is something wrong since we expect during deserialization, everything is empty ?

Reply:
```c++
template <typename Map, typename MappedDeserializer>
void deserializeTwoLevelHashMap(Map & map, MappedDeserializer && mapped_deserializer, Arena & pool, ReadBuffer & rb)
{
    deserializeHashMap<Map, MappedDeserializer>(map, std::move(mapped_deserializer), pool, rb);
    map.readUpdatedBuckets(rb); /// recover buckets updated status
}
```

For above impl,before `readUpdatedBuckets`, the `deserializeHashMap ` will reinsert all key, which will cause the `bucket_updated_flags ` all are true.